File: code/2025/04/05/98. 验证二叉搜索树.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
	bool isValidBST(TreeNode* r, long L = LONG_MIN, long R = LONG_MAX) {
		if (not r) {
			return true;
		}
		int x = r->val;
		return L < x and x < R and isValidBST(r->left, L, x) and isValidBST(r->right, x, R);
	}
};

File: code/2025/04/07/98. 验证二叉搜索树.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode], l = -inf, r = inf) -> bool:
        if (not root):
            return True
        x = root.val
        return l < x < r and self.isValidBST(root.left, l, x) and self.isValidBST(root.right, x, r)
